### ISSUE TITLE
Run on CPU?

### ISSUE THREAD

**Issue by barrykk1** [sb2702/free-ai-video-upscaler#12]
Is there any way to use the CPU?

**Comment by sb2702** [comment]
Hi,

I think it will work on the CPU when hardware acceleration is not available on some browsers (I think Chromium defaults to swiftshader), though it'd be a lot of work to re-architect the pipeline to run the models on the CPU.

Can you help me understand why CPU rendering would be preferred?

**Comment by barrykk1** [comment]
I use a Ryzen 7 5800XT + RX 580, I believe it would be faster because of the processor.

**Comment by sb2702** [comment]
I understand. It would be possible, you'd have to take the same [neural networks I trained](https://github.com/sb2702/websr/) and deploy them through Tensorflow and combine them to WASM or some other file.

I don't know if TensorflowJS or TFLite would introduce inefficiencies in a WASM environment but Tensorflow JS is incredibly inefficient compared to hard-coding neural networks in WebGPU/WebGL which is what I did:
https://free.upscaler.video/technical/architecture/

Also, even if your CPU is faster, it depends on where and how video encoding is structured in your browser. Typically browsers which have access to the GPU will use it .

For example, if you use chrome, you can take a look at chrome://gpu

If it says VideoEncoding and VideoDecoding is hardware accelerated, it might actually be doing the encoding/decoding on the GPU and so if that's the case, shuffling data back and forth between GPU and CPU would incur significant performance loss regardless of how powerful either the GPU or CPU is, because most of the actual runtime would be taken up by just shuffling data back and forth between devices, having a full pipeline on one device (all cpu or all GPU) is around 3x faster than a mixed pipeline per the results I found when testing:

https://webcodecsfundamentals.org/basics/rendering/


https://webcodecsfundamentals.org/concepts/cpu-vs-gpu/

https://medium.com/vectorly/building-a-more-efficient-background-segmentation-model-than-google-74ecd17392d5

I'm a little hesitant to spend time on what seems to be a niche use case, where the stated benefit may actually not be there, and which would actually require some effort to see if the benefit would be there

**Comment by barrykk1** [comment]
Okay, thank you for your attention.

**Comment by sb2702** [comment]
Sorry, as I mentioned, no plan to actually implement this given the amount of engineering it would take for the relatively speculative and niche benefit
